include/ginkgo/core/base/array.hpp

```diff
@@ -627,6 +655,41 @@ class copy_back_deleter<Array<T>> {
 };
 
 
+/**
+ * @internal
+ *
+ * Casts away const-ness from a array.
+ *
+ * This preserves is_owning() and re-wraps deleters, and can be used to
+ * implement functionality for classes that only store non-const arrays.
+ *
+ * @param array  the array to be cast. It will be empty with default deleter
+ *               afterwards.
+ * @returns a non-const array wrapping the same pointer and deleter on the same
+ *          executor as `array`.
+ */
+template <typename ValueType>
+Array<ValueType> array_const_cast(Array<const ValueType>&& array)
+{
+    auto deleter = array.data_.get_deleter();
+    std::function<void(const ValueType[])> new_deleter;
+    if (array.is_owning()) {
+        new_deleter =
+            typename Array<const ValueType>::default_deleter{array.exec_};
+    } else {
+        new_deleter = [deleter](auto ptr) { deleter(ptr); };
+    }
+    Array<ValueType> result{array.exec_, array.num_elems_,
+                            const_cast<ValueType*>(array.data_.release()),
+                            new_deleter};
+    array.num_elems_ = 0;
+    array.data_ = typename Array<const ValueType>::data_manager{
+        nullptr, typename Array<const ValueType>::default_deleter{array.exec_}};
```

Comment:
I think you can use `array.clear()` ? Then you can get rid of making this function a friend of `Array` as well

Reply:
In general yes, that's the same code, but I'm still not sure about this one, as `array.clear()` should assume an intact Array, while at this point with `data_.release()` we left the array in an incomplete state. Is that thinking too strictly in terms of invariants?

Reply:
The `friend` declaration will still be necessary to be able to steal the pointer via `release()`.

Reply:
I think with `data_.reset(nullptr)` in `clear` you would do the same that you do here ? But I agree you need the `friend` for the `release()`

Reply:
The main difference is that `data.reset(nullptr)` calls the deleter, while `release()` doesn't do anything - we need the latter behavior.